### ISSUE TITLE
Segfaults if second request is sent too quickly following first completion.

### ISSUE THREAD

**Issue by personsg** [ggerganov/llama.cpp#2303]
If a second request is sent too quickly after the first request finishes, the app crashes with `ggml_metal_get_buffer: error: buffer is nil`.

# Reproduction
I can reliably recreate either by
1. Sending a completion request
2. Waiting for response (with streaming on or off)
3. Immediately sending a second request.

Can also be recreated through the interface by
1. Sending a message
2. Quickly typing a second message into the box
3. Immediately hitting send as soon as it finishes streaming.

# Environment and Context

MacOS 13.4.1, M1 Max 64gb.
`./server -m ./models/7b/llama-2-7b.ggmlv3.q4_1.bin -ngl 1`

```
$ python3 --version
Python 3.10.8
$ make --version
GNU Make 3.81
$ g++ --version
14.0.0
```

**Comment by ridulfo** [comment]
I noticed this too when I was using llama.cpp for code completion with github copilot. Copilot was sending new requests before the previous one had finished.

The problem could be that the server tried to start a new completion before the old one is finished. 
Maybe the next request writes over the previous request's context? Is the mutex working properly?

I managed to stop it from happening by setting `svr.new_task_queue = [] { return new ThreadPool(1); };`
According to the [documentation](https://github.com/yhirose/cpp-httplib#default-thread-pool-support), httplib uses 12 threads by default.
